zebra-state/src/service/finalized_state/disk_db.rs

```diff
@@ -385,6 +405,15 @@ impl DiskDb {
     /// TODO: make private after the stop height check has moved to the syncer (#3442)
     ///       move shutting down the database to a blocking thread (#2188)
     pub(crate) fn shutdown(&mut self) {
+        if Arc::get_mut(&mut self.db).is_none() {
+            debug!(
+                "dropping cloned DiskDb, \
+                 but keeping shared database until the last reference is dropped",
+            );
+
+            return;
+        }
+
```

Comment:
`shutdown()` is also called manually by `commit_finalized_direct()` when set to stop at some height, so now it could do nothing if there is another reference. Is this an issue?

Reply:
Nice catch!

This might have been an issue when we're using `debug_stop_at_height`, it could have caused a hang. Temporary fix in 6a27bf4b4, permanent fix in #3442.

Reply:
There was also a potential race condition when checking if there was only one `Arc<DB>` clone. I fixed that by keeping an exclusive reference to the inner DB during cleanup.

(We still drop the reference when we return, but if it gets cloned again, that should just result in database errors.)